# Projekt ELECTRE III + PROMETHEE
Adam Gorgoń 145278
Jan Kabziński 145220

## Biblioteki

In [1]:
import pandas as pd

## Pobranie danych

In [2]:
data = pd.read_csv("ISWD - Arkusz1.csv", index_col="Nazwa")

In [3]:
data

Przekątna  RAM  Pamięć   
Nazwa                                                         
Apple iPhone 14 Pro 1TB                    6,1    6    1000  \
Samsung Galaxy S23 Ultra 12/1TB            6,8   12    1000   
Apple iPhone 14 Pro 256GB                  6,1    6     256   
Apple iPhone 14 Plus 512GB                 6,7    6     512   
ASUS ROG Phone 7 Ultimate 16/512GB        6,78   16     512   
Sony Xperia 1 V 12/256GB OLED 120Hz        6,5   12     256   
Apple iPhone 14 512GB                      6,1    6     512   
Huawei Mate 50 Pro 8/512GB                6,74    8     512   
Samsung Galaxy S23+ 8/512GB                6,6    8     512   
Apple iPhone 13 512GB                      6,1    4     512   
Nubia Redmagic 8 Pro 5G 16/512GB           6,8   16     512   
Apple iPhone 13 Mini 512GB                 5,4    4     512   
Sony Xperia 5 IV                           6,1    8     128   

                                      Częstotliwość [Hz]   
Nazwa                                                      
Apple iPhone 14 Pro 1TB                              120  \
Samsung Galaxy S23 Ultra 12/1TB                      120   
Apple iPhone 14 Pro 256GB                            120   
Apple iPhone 14 Plus 512GB                            60   
ASUS ROG Phone 7 Ultimate 16/512GB                   165   
Sony Xperia 1 V 12/256GB OLED 120Hz                  120   
Apple iPhone 14 512GB                                 60   
Huawei Mate 50 Pro 8/512GB                           120   
Samsung Galaxy S23+ 8/512GB                          120   
Apple iPhone 13 512GB                                 60   
Nubia Redmagic 8 Pro 5G 16/512GB                     120   
Apple iPhone 13 Mini 512GB                            60   
Sony Xperia 5 IV                                     120   

                                      Rozdzielczość głównego tylnego aparatu [Mpix]   
Nazwa                                                                                 
Apple iPhone 14 Pro 1TB                                                          48  \
Samsung Galaxy S23 Ultra 12/1TB                                                 200   
Apple iPhone 14 Pro 256GB                                                        48   
Apple iPhone 14 Plus 512GB                                                       12   
ASUS ROG Phone 7 Ultimate 16/512GB                                               13   
Sony Xperia 1 V 12/256GB OLED 120Hz                                              48   
Apple iPhone 14 512GB                                                            12   
Huawei Mate 50 Pro 8/512GB                                                       50   
Samsung Galaxy S23+ 8/512GB                                                      50   
Apple iPhone 13 512GB                                                            12   
Nubia Redmagic 8 Pro 5G 16/512GB                                                 50   
Apple iPhone 13 Mini 512GB                                                       12   
Sony Xperia 5 IV                                                                 12   

                                      Cena  
Nazwa                                       
Apple iPhone 14 Pro 1TB               8499  
Samsung Galaxy S23 Ultra 12/1TB       8699  
Apple iPhone 14 Pro 256GB             6949  
Apple iPhone 14 Plus 512GB            6899  
ASUS ROG Phone 7 Ultimate 16/512GB    6799  
Sony Xperia 1 V 12/256GB OLED 120Hz   6499  
Apple iPhone 14 512GB                 6499  
Huawei Mate 50 Pro 8/512GB            6499  
Samsung Galaxy S23+ 8/512GB           5799  
Apple iPhone 13 512GB                 5699  
Nubia Redmagic 8 Pro 5G 16/512GB      5299  
Apple iPhone 13 Mini 512GB            4999  
Sony Xperia 5 IV                      4899

In [4]:
data.dtypes

Przekątna                                        object
RAM                                               int64
Pamięć                                            int64
Częstotliwość [Hz]                                int64
Rozdzielczość głównego tylnego aparatu [Mpix]     int64
Cena                                              int64
dtype: object

In [5]:
data["Przekątna"] = data["Przekątna"].str.replace(',', '.').astype(float)

## Informacje o zbiorze danych

1. Jaka jest domena problemu?
Analiza najnowszych telefonów.
2. Jakie jest źródło danych?
Ceny telefonów z sklepu internetowego x-kom z czerwca 2023 roku. Dane zostały zebrane przeze mnie.
3. Jaki jest punkt widzenia decydenta?
Pomoc w wyborze najbardziej opłacalnego telefonu z najwyższej półki cenowej.
4. Jaka jest liczba rozważanych wariantów? Czy w oryginalnym zbiorze danych było ich więcej?
Jest rozważane 13 wariantów. Sam tworzyłem zbiór więc nie było więcej. W źródle, czyli w sklepie internetowym, można znaleźć wiele więcej wariantów.

In [6]:
warianty = data.index.tolist()
len(warianty)

13

5. Opisz jeden z rozważanych wariantów - jak jest oceniony na poszczególnych kryteriach, jak dobry jest
na tle innych.
Samsung Galaxy S23 Ultra 12/1TB. Jest to telefon z największą przekątną, pamięcią oraz rozdzielczość głównego tylnego aparatu. Ma również największą cene ze wszystkich.

In [7]:
data.loc["Samsung Galaxy S23 Ultra 12/1TB"]

Przekątna                                           6.8
RAM                                                12.0
Pamięć                                           1000.0
Częstotliwość [Hz]                                120.0
Rozdzielczość głównego tylnego aparatu [Mpix]     200.0
Cena                                             8699.0
Name: Samsung Galaxy S23 Ultra 12/1TB, dtype: float64

6. Ile rozwazasz kryteriów? Czy było ich więcej w oryginalnym zbiorze danych?
Rozważamy 6 kryteriów. Początkowo chcieliśmy dodać pojemność baterii i rozdzielczość trzech aparatów ale nie każdy wariant miał informacje związane z nimi. Były też informacje o wymiarach telefonu, które uznaliśmy za bezużyteczne.
7. Jakie są typy poszczególnych kryteriów? Scharakteryzuj je pod kątem:
* przekątna - ciągła, zysk, [5.4, 6.8]
* RAM - dystkretna, zysk, [4, 16]
* pamięć - dyskretna, zysk [128, 1000]
* częstotliwość - ciągła, zysk [60, 165]
* rozdzielczość głównego tylnego aparatu - ciągła, zysk [12, 200]
* cena - ciągła, koszt, [4899, 8699]

In [8]:
kryteria = data.columns.tolist()
kryteria

['Przekątna',
 'RAM',
 'Pamięć',
 'Częstotliwość [Hz]',
 'Rozdzielczość głównego tylnego aparatu [Mpix]',
 'Cena']

In [9]:
skale = pd.merge(data.min().rename('min'), data.max().rename("max"), left_index=True, right_index=True).loc[kryteria]
skale

min     max
Przekątna                                         5.4     6.8
RAM                                               4.0    16.0
Pamięć                                          128.0  1000.0
Częstotliwość [Hz]                               60.0   165.0
Rozdzielczość głównego tylnego aparatu [Mpix]    12.0   200.0
Cena                                           4899.0  8699.0

In [10]:
typy = {kryterium: typ for kryterium, typ in zip(kryteria, ["koszt"] * 5 + ["zysk"])}
typy

{'Przekątna': 'koszt',
 'RAM': 'koszt',
 'Pamięć': 'koszt',
 'Częstotliwość [Hz]': 'koszt',
 'Rozdzielczość głównego tylnego aparatu [Mpix]': 'koszt',
 'Cena': 'zysk'}

8. Jakie są wagi poszczególnych kryteriów?
* przekątna - 7
* RAM - 6
* pamięć - 8
* częstotliwość - 3
* rozdzielczość głównego tylnego aparatu - 3
* cena - 10

In [11]:
wagi = {kryterium: waga for kryterium, waga in zip(kryteria, (7, 6, 8, 3, 3, 10))}
wagi

{'Przekątna': 7,
 'RAM': 6,
 'Pamięć': 8,
 'Częstotliwość [Hz]': 3,
 'Rozdzielczość głównego tylnego aparatu [Mpix]': 3,
 'Cena': 10}

9. Czy istnieją jakieś dominacje w zbiorze wariantów? Jeśli tak, podaj je.
Istnieją dominacje.
* Nubia Redmagic 8 Pro 5G 16/512GB dominuje: Apple iPhone 13 512GB, Samsung Galaxy S23+ 8/512GB, Sony Xperia 1 V 12/256GB OLED 120Hz, Apple iPhone 14 512GB, Huawei Mate 50 Pro 8/512GB, ASUS ROG Phone 7 Ultimate 16/512GB, Apple iPhone 14 Plus 512GB, Apple iPhone 14 Pro 256GB.
* Huawei Mate 50 Pro 8/512GB dominuje Apple iPhone 14 512GB.
* ASUS ROG Phone 7 Ultimate 16/512GB dominuje Apple iPhone 14 Plus 512GB.
10. Jak teoretycznie powinien prezentować się najlepszy wariant w Twoim zbiorze danych?
Najlepszy wariant powinien mieć dużą przekątną i pamięć z racjonalną ceną.
11. Spośród rozważanych wariantów, który wydaje się być najlepszy/lepszy od innych i dlaczego?
Nubia Redmagic 8 Pro 5G 16/512GB. Na oko wydaje się, że ma najlepsze kryteria i dominuje wiele innych wariantów.
12. Sposród rozważanych wariantów, który wydaje się być najsłabszy/ znacząco słabszy od innych i dlaczego?
Apple iPhone 14 Pro 1TB. Ma dużo gorsze kryteria od Samsung Galaxy S23 Ultra 12/1TB, a jest tylko 200 złotych tańszy.

## Analiza problemu z wykorzystaniem ELECTRE III

### Informacje preferencyjne podane na wejście metody.

In [12]:
qs = {kryterium: q for kryterium, q in zip(kryteria, (0.1, 2, 200, 50, 5, 200))}
ps = {kryterium: p for kryterium, p in zip(kryteria, (0.3, 4, 250, 70, 20, 1000))}
vs = {kryterium: v for kryterium, v in zip(kryteria, (1, 10, 500, 100, 100, 2000))}

### Obliczanie globalnego współczynnika zgodności

In [13]:
zgodnosci = {}
for kryterium in kryteria:
    zgodnosc = []
    for wariant1 in warianty:
        wiersz = []
        for wariant2 in warianty:
            a_value = data.loc[wariant1, kryterium]
            b_value = data.loc[wariant2, kryterium]
            q = qs.get(kryterium)
            p = ps.get(kryterium)
            v = vs.get(kryterium)
            if typy.get(kryterium) == "zysk":
                if a_value - b_value <= q:
                    wiersz.append(1.0)
                elif a_value - b_value > p:
                    wiersz.append(0.0)
                else:
                    wiersz.append((b_value - a_value + p) / (p - q))
            else:
                if b_value - a_value <= q:
                    wiersz.append(1.0)
                elif b_value - a_value > p:
                    wiersz.append(0.0)
                else:
                    wiersz.append((p - (b_value - a_value)) / (p - q))
        zgodnosc.append(wiersz)
    zgodnosci[kryterium] = pd.DataFrame(zgodnosc, index=warianty, columns=warianty)

In [14]:
C = []
for wariant1 in warianty:
    wiersz = []
    for wariant2 in warianty:
        suma = 0
        for kryterium in kryteria:
            suma += wagi.get(kryterium) * zgodnosci.get(kryterium).loc[wariant1, wariant2]
        wiersz.append(suma/sum(wagi.values()))
    C.append(wiersz)
C = pd.DataFrame(C, index=warianty, columns=warianty)

In [15]:
C

Apple iPhone 14 Pro 1TB   
Apple iPhone 14 Pro 1TB                              1.000000  \
Samsung Galaxy S23 Ultra 12/1TB                      1.000000   
Apple iPhone 14 Pro 256GB                            0.783784   
Apple iPhone 14 Plus 512GB                           0.662162   
ASUS ROG Phone 7 Ultimate 16/512GB                   0.702703   
Sony Xperia 1 V 12/256GB OLED 120Hz                  0.783784   
Apple iPhone 14 512GB                                0.662162   
Huawei Mate 50 Pro 8/512GB                           0.783784   
Samsung Galaxy S23+ 8/512GB                          0.783784   
Apple iPhone 13 512GB                                0.662162   
Nubia Redmagic 8 Pro 5G 16/512GB                     0.783784   
Apple iPhone 13 Mini 512GB                           0.472973   
Sony Xperia 5 IV                                     0.702703   

                                      Samsung Galaxy S23 Ultra 12/1TB   
Apple iPhone 14 Pro 1TB                                      0.567568  \
Samsung Galaxy S23 Ultra 12/1TB                              1.000000   
Apple iPhone 14 Pro 256GB                                    0.351351   
Apple iPhone 14 Plus 512GB                                   0.500000   
ASUS ROG Phone 7 Ultimate 16/512GB                           0.702703   
Sony Xperia 1 V 12/256GB OLED 120Hz                          0.513514   
Apple iPhone 14 512GB                                        0.310811   
Huawei Mate 50 Pro 8/512GB                                   0.540541   
Samsung Galaxy S23+ 8/512GB                                  0.445946   
Apple iPhone 13 512GB                                        0.310811   
Nubia Redmagic 8 Pro 5G 16/512GB                             0.702703   
Apple iPhone 13 Mini 512GB                                   0.310811   
Sony Xperia 5 IV                                             0.351351   

                                      Apple iPhone 14 Pro 256GB   
Apple iPhone 14 Pro 1TB                                0.729730  \
Samsung Galaxy S23 Ultra 12/1TB                        0.729730   
Apple iPhone 14 Pro 256GB                              1.000000   
Apple iPhone 14 Plus 512GB                             0.878378   
ASUS ROG Phone 7 Ultimate 16/512GB                     0.918919   
Sony Xperia 1 V 12/256GB OLED 120Hz                    1.000000   
Apple iPhone 14 512GB                                  0.878378   
Huawei Mate 50 Pro 8/512GB                             1.000000   
Samsung Galaxy S23+ 8/512GB                            1.000000   
Apple iPhone 13 512GB                                  0.878378   
Nubia Redmagic 8 Pro 5G 16/512GB                       1.000000   
Apple iPhone 13 Mini 512GB                             0.689189   
Sony Xperia 5 IV                                       0.918919   

                                      Apple iPhone 14 Plus 512GB   
Apple iPhone 14 Pro 1TB                                 0.540541  \
Samsung Galaxy S23 Ultra 12/1TB                         0.729730   
Apple iPhone 14 Pro 256GB                               0.594595   
Apple iPhone 14 Plus 512GB                              1.000000   
ASUS ROG Phone 7 Ultimate 16/512GB                      1.000000   
Sony Xperia 1 V 12/256GB OLED 120Hz                     0.689189   
Apple iPhone 14 512GB                                   0.810811   
Huawei Mate 50 Pro 8/512GB                              1.000000   
Samsung Galaxy S23+ 8/512GB                             1.000000   
Apple iPhone 13 512GB                                   0.810811   
Nubia Redmagic 8 Pro 5G 16/512GB                        1.000000   
Apple iPhone 13 Mini 512GB                              0.810811   
Sony Xperia 5 IV                                        0.594595   

                                      ASUS ROG Phone 7 Ultimate 16/512GB   
Apple iPhone 14 Pro 1TB                                         0.378378  \
Samsung Galaxy S23 Ultra 12/1TB                             

### Test zgodności

In [16]:
(C <= 1).all(axis=None) and (C >= 0).all(axis=None)

True